### ISSUE TITLE
DEADLOCK when invoke is indirectly called again in subprocess chain

### ISSUE THREAD

**Issue by jenisys** [pyinvoke/invoke#318]
VERSION: invoke-0.12.0 (note: invoke-0.11.1 does not have this problem)
PLATFORM: OSX 10.11 (El Capitan)

I ran into a weird problem when I run invoke again by another subprocess.
The second invoke process (tasks2.py) never seems to terminate (only when you kill the process-group)
after first analysis. When I run the same without the first invoke, everything is fine.
NOTE: some-subprocess2.py is a simplistic python script with 2 lines (print+sys.exit).

```
PROCESS-TREE:
+-- invoke tasks.py: task run("some-subprocess.py")
|   +--  some-subprocess.py: subprocess.call("invoke ...") # in another directory
|      +-- invoke tasks2.py: run("some-subproces2.py")     # DEADLOCK somewhere here
```

**Comment by jenisys** [comment]
After further analysis, the most inner subprocess "some-subprocess2.py" terminates with ``sys.exit(0)`` (or other exit-codes), but invoke-2 subprocess never seems to capture this process-termination (even when multi-seconds sleeps are used in "some-subprocess2.py"). Note that "some-subprocess2.py" prints something to stdout (but not to stderr).

```text
INVOKE-2 CALLGRAPH:
  Thread 1 (main-thread)
    + 2821 Py_Main  (in Python) + ...
        ... (many python related C-API calls, like: PyObject_Call, etc.)
        + 2821 PyThread_acquire_lock  (in Python) + 116  [0x100101994]
           + 2821 _pthread_cond_wait (DEADLOCKED HERE)
  
  Thread 2
     ...
     + 2821 file_read  (in Python) + 246  [0x100033fa6]
       + 2821 fread  (in libsystem_c.dylib) + 48  [0x7fff96151e0e]
         + 2821 __fread  (in libsystem_c.dylib) + 393  [0x7fff96151fae]
           + 2821 __srefill1  (in libsystem_c.dylib) + 24  [0x7fff96157771]
             + 2821 _sread  (in libsystem_c.dylib) + 16  [0x7fff9615815d]
               + 2821 __read_nocancel  (in libsystem_kernel.dylib) + 10  [0x7fff93fa9086]  <== NEVER-ENDS
```

**Comment by bitprophet** [comment]
0.12.1 should have fixed this, I believe (see #308). Please comment if not the case!

**Comment by jenisys** [comment]
@bitprophet 
FIXED in invoke-0.12.2
Thanks :smile:

**Comment by jenisys** [comment]
CORRECTION: The DEADLOCK is fixed with invoke-0.12.2, but I need to apply the patch for issue #317 (merge-dicts problem). Otherwise, I cannot run any task. Sorry.
